geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java

```diff
@@ -0,0 +1,715 @@
+        accessor.invokeAsync(() -> executeGets(entries, workSeconds)),
+        accessor.invokeAsync(() -> executeRemoves(entries, workSeconds)));
+
+    // Retry the clear operation on the region until success (server2 will go down, but other
+    // members will become primary for those buckets previously hosted by server2).
+    executeClearWithRetry(getVM(coordinatorVM.vmNumber));
+
+    // Wait for member to get back online.
+    server2.invoke(() -> {
+      cacheRule.createCache();
+      initDataStore(RegionShortcut.PARTITION_REDUNDANT);
+      await().untilAsserted(
+          () -> assertThat(InternalDistributedSystem.getConnectedInstance()).isNotNull());
+    });
+
+    // Let asyncInvocations finish.
+    for (AsyncInvocation<Void> asyncInvocation : asyncInvocationList) {
+      asyncInvocation.await();
+    }
+
+    // Assert Region Buckets are consistent.
+    asList(accessor, server1, server2).forEach(vm -> vm.invoke(this::waitForSilence));
+    accessor.invoke(this::assertRegionBucketsConsistency);
+  }
+
+  /**
+   * The test does the following (clear coordinator is chosen through parameters):
+   * - Populates the Partition Region.
+   * - Verifies that the entries are synchronized on all members.
+   * - Sets the {@link MemberKiller} as a {@link CacheWriter} to stop a non-coordinator VM while the
+   * clear is in progress (the member has primary buckets, though, so participates on
+   * the clear operation).
+   * - Launches two threads per VM to continuously execute gets, puts and removes for a given time.
+   * - Clears the Partition Region (at this point the non-coordinator is restarted).
+   * - Asserts that the clear operation failed with PartitionedRegionPartialClearException (primary
+   * buckets on the the restarted members are not available).
+   */
+  @Test
+  @Parameters(method = "coordinators")
+  @TestCaseName("[{index}] {method}(Coordinator:{0})")
+  public void clearOnNonRedundantPartitionRegionWithConcurrentPutGetRemoveShouldFailWhenNonCoordinatorMembersAreBounced(
+      TestVM coordinatorVM) throws InterruptedException {
+    final int entries = 7500;
+    final int workSeconds = 45;
+    parametrizedSetup(RegionShortcut.PARTITION);
+    registerVMKillerAsCacheWriter(Collections.singletonList(server2));
+    populateRegion(accessor, entries, asList(accessor, server1, server2));
+
+    // Let all VMs (except the one to kill) continuously execute gets, put and removes for 45
+    // seconds.
+    List<AsyncInvocation<Void>> asyncInvocationList = Arrays.asList(
+        server1.invokeAsync(() -> executeGets(entries, workSeconds)),
+        server1.invokeAsync(() -> executePuts(entries, workSeconds)),
+        accessor.invokeAsync(() -> executeGets(entries, workSeconds)),
+        accessor.invokeAsync(() -> executeRemoves(entries, workSeconds)));
+
+    // Clear the region.
+    getVM(coordinatorVM.vmNumber).invoke(() -> {
+      assertThatThrownBy(() -> cacheRule.getCache().getRegion(REGION_NAME).clear())
```

Comment:
The PartialClearException could happen here but not 100% will happen. Your assertion could fail here.

Reply:
@gesterzhou 

There are three members in this test: `ACCESSOR`, `SERVER1` and `SERVER2`, both servers host primary buckets locally for the region. The `clear` operation is executed either from the `ACCESSOR` or `SERVER1`, and `SERVER2` is always killed through the `MemberKiller` class (`CacheWriter.beforeRegionClear()`) before the `clear` is locally executed.
According to what I understand, the `PartitionedRegionPartialClearException` should always be thrown as `SERVER2` doesn't get started (after the initial shutdown) until the `clear` "finishes", meaning that the clear coordinator **_will never be able to get the lock for all the primary buckets_** (the region is configured as `PARTITION`, not as `PARTITION_REDUNDANT`, so there are no secondary buckets that can become primary while `SERVER2` is down).
Am I missing something here?, shouldn't the exception be thrown no matter what, given the above conditions?.

---

As a side note, the test `clearOnNonRedundantPartitionRegionWithConcurrentPutGetRemoveShouldFailWhenNonCoordinatorMembersAreBounced` failed 4 times out of 100 executions, the actual failure depends on whether the `clear` is executed on the `ACCESSOR` or `SERVER1` vm.
The two failures seen when the `clear` is executed on the the `ACCESSOR` node are basically assertion errors, caused by the fact the the `PartitionedRegionPartialClearException` is not thrown (discussed above).
The two failures seen when the `clear` is executed on `SERVER1`, on the other hand, are caused by the fact that the exception message is different from the one expected by the test (I'll change this and update the `PR`).

Reply:
@gesterzhou 
I've modified the test assertion to always expect the `PartitionedRegionPartialClearException` but don't check the actual message, all the related failures are gone now.
The question still remains, though, the `PartitionedRegionPartialClearException` is always thrown under the conditions set by the test (that's what I would expect), why do you think the mentioned exception shouldn't be thrown 100% of time under these conditions?.

Reply:
Only when the primary holder cannot be found any where, the PartitionedRegionPartialClearException will be throw. 

In your case, only killing server2. There're chances and races that the primary buckets are recreated quick enough on server1. So PartitionedRegionPartialClearException is not 100%.

Reply:
Maybe I'm missing something here, but there's no activity on the cluster at the time `server2` is killed, so the primaries previously hosted by that server shouldn't be recreated anywhere (and there are no secondaries to promote as the region is just `PARTITION`, not `REDUNDANT_PARTITION`).